s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -856,6 +834,21 @@ import scala.util.{Failure, Success, Try}
     chunkAndRequest(s3Location, contentType, s3Headers, chunkSize)(chunkingParallelism)
       .toMat(completionSink(s3Location, s3Headers.serverSideEncryption))(Keep.right)
 
+  /**
+   * Uploads a stream of ByteStrings along with a context to a specified location as a multipart upload. The
+   * chunkUploadSink parameter allows you to act upon the context when a chunk has been uploaded to S3.
+   */
+  def multipartUploadWithContext[C](
+      s3Location: S3Location,
+      chunkUploadSink: Sink[(UploadPartResponse, immutable.Iterable[C]), NotUsed],
+      contentType: ContentType = ContentTypes.`application/octet-stream`,
+      s3Headers: S3Headers,
+      chunkSize: Int = MinChunkSize,
+      chunkingParallelism: Int = 4
+  ): Sink[(ByteString, C), Future[MultipartUploadResult]] =
```

Comment:
Given that it returns a sink and accepts a sink, and will feed elements through one side to the other, should it perhaps rather return a `Flow[(ByteString, C), (UploadPartResponse, Iterable[C]), Future[MultipartUploadResult]]`, which expresses that more clearly, and is materialized as one stream, or am I missing something that makes it specifically require sink + sink?

Reply:
I can see both sides, from one end a `Flow` from a technically makes sense for the reasons you stated however from a usability point of view its slightly odd because you will always use it as a `Sink` and hence you will end up doing `.via(multipartUploadWithContext(...)).to(Sink.ignore)` all of the time rather than just `.to(multipartUploadWithContext(...))`.

I spoke with a colleague about this and he said that as far as he could recall there were already existing examples in Alpakka where a `Flow` is used so I will do some quick research and if its the case then I will change it for the sake of consistency.

Reply:
Yeah, I think that's how the other connectors in general look. A flow can also be run like so `source.via(flow).run()` (short for `.runWith(Sink.ignore)`, so the syntax difference vs `source.runWith(sink)` is not huge. 

I did expect that you want to run this with something consuming those contexts though, isn't that kind of the point?

Reply:
True, but in my case (I actually adjusted my code to work with this PR) the "callback" is a `Sink` (actually `Committer.sink(committerSettings)` from Alpakka Kafka) and since the current implementation is keeping the return type of `multipartUploadWithContext` as a `Sink` that part is unchanged (I just had to adjust `contramap` to pass in the `CursorContext`).

So if I changed it to a `Flow[(ByteString, C), (UploadPartResponse, Iterable[C]), Future[MultipartUploadResult]]` I would have do more adjustments (i.e. using `Sink`'s within a flow) but as you said this is conventional for this usecase.

I will do change to `Flow[(ByteString, C), (UploadPartResponse, Iterable[C]), Future[MultipartUploadResult]]` locally and update my PR with those changes to see if there are any problems (there shouldn't be) and I will also do some future testing

Reply:
So I tried changing it to `Flow[(ByteString, C), (UploadPartResponse, Iterable[C]), Future[MultipartUploadResult]]` and I realized that this is the wrong type signature for what is being done.

I will explain this the best way I can, but the situation here is different to what is done with other Alpakka connectors. For other Alpakka modules, there are "lower" level building blocks (composed of `Source`/`Flow`/`Sink`) which are then combined by the user of their API. On the other hand specifically with S3's `multipartUpload`/`resumeMultipartUpload` the implementation is the opposite, i.e. this "lower" level building block composition is abstracted away in the API (in `chunkAndRequest`/`chunkAndRequestWithContext`) and for good reason (this is unnecessary complication that the user doesn't need to know).

The other issue with `Flow[(ByteString, C), (UploadPartResponse, Iterable[C]), Future[MultipartUploadResult]]` is that its saying the output of the flow is `[(UploadPartResponse, Iterable[C])` which is not correct, this is actually an input (i.e. the user has to define how to act on `(UploadPartResponse, immutable.Iterable[C])`, this is not something that is implemented in the Alpakka S3 flow.

To simplify it down further, what we are actually doing is interposing one sink (the chunkUploadSink) inside of another sink (the s3UploadSink) with the method of interposing being an implementation detail of s3UploadSink (since the implementation of s3UploadSink itself hidden as an implementation detail).

Ultimately I think we have 3 ways of solving this feature

1. This current implementation, where the user has to provide an additional `Sink[(UploadPartResponse, immutable.Iterable[C]), NotUsed]` as a parameter.
2. What you suggested at https://github.com/akka/alpakka/issues/2760#issuecomment-966064402 which is exposing the internals, this means that a lot of the internal implementation of `S3Stream.scala` would become public so that users are able to combose the multipart upload themselves how they want. In other words users would actually write https://github.com/akka/alpakka/pull/2770/commits/3aec0df850768bf8c015ad8e71e7a19636929c26#diff-cbe9040cf5467cee6ffdbd2896491e2da2acea2aec409e9075c5cb15e3180e59R1227-R1261 themselves (and by association any of the functions in that method would need to be public).
3. Another 3rd alternative which I just came up with is that rather than `multipartUpload`/`resumeMultipartUpload` only returning a `Sink[ByteString, Future[MultipartUploadResult]]`, it would instead return a `(Sink[ByteString, Future[MultipartUploadResult]], Source[String])`. The `Source[String]` is then a reference which you can pass into a second function `chunkResponseSink(uploadId: String): Sink[(UploadPartResponse, immutable.Iterable[C]), NotUsed]` as an `uploadId` which you would call right after calling `multipartUpload`/`resumeMultipartUpload`. The issue with this approach is that in order to not break the public API of `multipartUpload`/`resumeMultipartUpload` we would have to make new versions and secondly we would have to store some additional state within `multipartUpload`/`resumeMultipartUpload` since we have to hold onto an `uploadId` reference.

Reply:
Thanks for investigating that a bit. Since it is anyway a bit of a power-user feature/API I think the current sink-sink API is ok.

Reply:
Perfect thanks, I will work on adding tests to finalize the PR. Do you think its realistic to have this out before the next release (i.e. https://github.com/akka/alpakka/issues/2775) if I finish it in the next couple of days or is this pushing it too much?

Reply:
I wouldn't want make it a promise but we can try, wdyt @ennru ?

Reply:
We can wait a few more days before kicking off the release.